### ISSUE TITLE
Bug in floating-point conversions

### ISSUE THREAD

**Issue by LeventErkok** [Z3Prover/z3#1564]
@wintersteiger 

Hi Christoph,

I noticed you made some recent changes to the floating-point theory recently, squashing some bugs. Unfortunately, I think this introduced a bug with conversions. Consider the following:

```
(set-option :produce-models true)
(set-logic QF_FPBV)
(define-fun s1 () (_ FloatingPoint 11 53) ((_ to_fp 11 53) roundNearestTiesToEven (/ 1.0 2.0)))
(define-fun s3 () (_ BitVec 64) #x0000000000000000)
(define-fun s8 () RoundingMode roundNearestTiesToEven)
(declare-fun s0 () (_ FloatingPoint 11 53)) ; tracks user variable "x"
(define-fun s2 () Bool (fp.eq s0 s1))
(define-fun s4 () Bool (fp.isNaN s0))
(define-fun s5 () Bool (fp.isInfinite s0))
(define-fun s6 () Bool (or s4 s5))
(define-fun s7 () Bool (not s6))
(define-fun s9 () (_ BitVec 64) ((_ fp.to_sbv 64) s8 s0))
(define-fun s10 () (_ BitVec 64) (ite s7 s9 s3))
(define-fun s11 () Bool (= s3 s10))
(assert s2)
(assert (not s11))
(check-sat)
(get-value (s0))
```

With a freshly built z3, we get:

```
sat
((s0 (fp #b0 #b01111111110 #x0000000000000)))
```

I believe this model is bogus. With a few days old z3, I get:

```
unsat
```

The benchmark is indeed unsatisfiable, and the model found by the latest z3 is bogus. (Essentially, it's proving that rounding down `0.5` from `Double` to a 64-bit value must produce `0`.)

**Comment by wintersteiger** [comment]
Took another look: `((_ fp.to_sbv 64) s8 s0)` rounds `0.5` to signed BV with rounding mode `RNE` and the result is `0`. I think this is correct (`0` is even) and if, e.g. you replace `RNE` with `RNA` you get `1`, which is also correct.

**Comment by wintersteiger** [comment]
Aha, but there's a constraint says the output shouldn't be zero. It's a bug in the handling of underspecified cases of `fp.to_sbv`; for some reason it thinks this case is not specified.

**Comment by LeventErkok** [comment]
Hi Christoph.. Thanks for the analysis!

(Regarding the original report, which I changed: You are correct: I managed to confuse myself. There was a test suite failure, and I managed to pick the actually passing test from Double to Integer conversion via reals, which always produced `unknown`. It's actually constant propagation even in that case, but I understand that it's not quite well supported.)

Regarding the present bug: I'm pretty sure this got introduced in the last few days. It was passing in my Travis-CI environment just a few days ago and started failing sometime Friday. So, whatever the issue is, it must've been introduced in the last few commits. I hope that helps it narrow it down!
